### PR TITLE
[ttl] Thread local DFBs

### LINKED ISSUE #474: [external bug report][from claude] ttl.mul on multi-tile tensors fails to legalize when fused with reduce_sum
## Description

`reduce_sum(a * b, scaler, dims=[0, 1])` where `a` and `b` are `(1, 4)` tile DFB blocks fails during compilation with:

```
error: unknown: failed to legalize operation 'ttl.mul' that was explicitly marked illegal
```

The `ttl.mul` on `tensor<1x4x!ttcore.tile<32x32, bf16>>` operands is generated as an intermediate feeding into `ttl.reduce`, but the legalization pass does not handle multi-tile elementwise ops fused into reductions.

The same pattern works fine when the DFB blocks are `(1, 1)` tiles.

## Reproducer

```python
import torch
import ttl
import ttnn

TILE = 32
HEAD_TILES = 4

@ttl.operation(grid=(1, 1))
def repro_kernel(inp_a, inp_b, scaler, out):
    a_dfb = ttl.make_dataflow_buffer_like(inp_a, shape=(1, HEAD_TILES), buffer_factor=2)
    b_dfb = ttl.make_dataflow_buffer_like(inp_b, shape=(1, HEAD_TILES), buffer_factor=2)
    sc_dfb = ttl.make_dataflow_buffer_like(scaler, shape=(1, 1), buffer_factor=1)
    out_dfb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), buffer_factor=2)

    @ttl.compute()
    def compute():
        with sc_dfb.wait() as sc:
            with a_dfb.wait() as av, b_dfb.wait() as bv:
                with out_dfb.reserve() as o:
                    o.store(ttl.math.reduce_sum(av * bv, sc, dims=[0, 1]))

    @ttl.datamovement()
    def dm_read():
        with sc_dfb.reserve() as blk:
            tx = ttl.copy(scaler[0, 0], blk); tx.wait()
        with a_dfb.reserve() as blk:
            tx = ttl.copy(inp_a[0:1, 0:HEAD_TILES], blk); tx.wait()
        with b_dfb.reserve() as blk:
            tx = ttl.copy(inp_b[0:1, 0:HEAD_TILES], blk); tx.wait()

    @ttl.datamovement()
    def dm_write():
        with out_dfb.wait() as blk:
            tx = ttl.copy(blk, out[0, 0]); tx.wait()

if __name__ == "__main__":
    device = ttnn.open_device(device_id=0)
    try:
        hd = HEAD_TILES * TILE
        a = ttnn.from_torch(torch.randn(TILE, hd, dtype=torch.bfloat16),
                            dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                            device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
        b = ttnn.from_torch(torch.randn(TILE, hd, dtype=torch.bfloat16),
                            dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                            device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
        scaler = ttnn.from_torch(torch.ones(TILE, TILE, dtype=torch.bfloat16),
                                 dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                                 device=device, memory_config=ttnn.L1_MEMORY_CONFIG)
        out = ttnn.from_torch(torch.zeros(TILE, TILE, dtype=torch.bfloat16),
                              dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                              device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
        repro_kernel(a, b, scaler, out)
    finally:
        ttnn.close_device(device)
```

## Relevant MLIR (initial)

The offending IR in compute:

```mlir
%12 = ttl.mul %7, %9 : tensor<1x4x!ttcore.tile<32x32, bf16>>,
                        tensor<1x4x!ttcore.tile<32x32, bf16>>
                     -> tensor<1x4x!ttcore.tile<32x32, bf16>>
%13 = ttl.reduce %12, %5 0 : i32 [0, 1]
      : (tensor<1x4x!ttcore.tile<32x32, bf16>>,
         tensor<1x1x!ttcore.tile<32x32, bf16>>)
      -> tensor<1x1x!ttcore.tile<32x32, bf16>>
```

## Environment

- ttlang 0.1.7.dev58
- Wormhole B0 (BH)
- Tested on HW via run-test.sh

### PR DESCRIPTION
**Pull request by brnorris03**
Automatically insert intermediate dataflow buffers when a fused expression chain feeds into an op whose tile-level lowering reads from the unpacker (reduce, bcast, matmul, transpose). Previously the user had to manually declare and wire these buffers; now the compiler materializes them.

A new `DFBInputOpInterface` declares which operands of a tensor-level op require DFB-attached values. The `ttl-insert-intermediate-dfbs` pass walks all ops implementing this interface and, for each non-DFB-attached operand, inserts a scratch DFB with `bind_cb`, `cb_reserve`, `store`, `cb_wait`, and `attach_cb`. The `ttl-insert-cb-sync` pass then fills in the matching `cb_push` and `cb_pop`.

Scratch DFBs use `block_count=2` because the hardware requires double-buffering when the packer and unpacker operate on the same CB within one thread. They are tagged with `ttl.compiler_allocated` so that `ttl-finalize-dfb-indices` can collect them into a `ttl.scratch_dfbs` module attribute. The Python runtime reads this attribute after compilation and builds the corresponding `CBDescriptor` entries for `ttnn.generic_op`.

The `ttl-insert-cb-sync` push-placement logic is extended to bound the search by the next `cb_wait` on the same CB, not just the next `cb_reserve`. Without this, `cb_push` for intra-thread DFBs was placed after `cb_pop`, causing a deadlock.

Pipeline order: `ttl-insert-intermediate-dfbs` runs first (creates scratch DFBs), then `ttl-insert-cb-sync` (inserts push/pop for all DFBs including scratch), then `convert-ttl-to-compute` (sees all operands as DFB-attached).

Tested with add->reduce, matmul->reduce, and a multinode per-tile softmax kernel where the user declares only I/O and scaler DFBs and the compiler creates three intermediates for `reduce_max`, `exp`, and `reduce_sum` results.

Not in this PR: 
- DFB index reuse. Currently each scratch DFB gets a fresh index, which can lead to unnecessary L1 pressure when many intermediates are created. A follow-up PR will implement linear scan allocation to reuse indices when lifetimes do not overlap.
- Result materialization (e.g., DST result to DFB-input op), #508

Closes #474

### CODE REVIEW

**Review by zoecarver — APPROVED** [review]
LGTM!

File: test/python/test_compiler_allocated_dfb.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
+# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
```
Comment (zoecarver) [inline test/python/test_compiler_allocated_dfb.py:1]: A few test patterns that would be good to check:
* thread local/implicit dfb inside loop/inside a few interesting/deep control flows
* thread local dfb used with implicitly pop'ed dfb

**Comment by zoecarver** [comment]
[test_gdn_kernel.py](https://github.com/user-attachments/files/26760699/test_gdn_kernel.py)

Tried using this for a gdn kernel and the above program crashed:
```
virtual void mlir::RewriterBase::replaceOp(Operation *,
     ValueRange): Assertion `op->getNumResults() == newValues.size() && "incorrect # of replacement values"' failed.
```

**Comment by brnorris03** [comment]
>[test_gdn_kernel.py](https://github.com/user-attachments/files/26760699/test_gdn_kernel.py)

Fixed assertion and added issue to implement multi-output compute properly in the relevant pass in a subsequent PR (should be done quickly).
